Fixes #17303 - correct position of validation msg for parameters

kgaikwad opened the pull request:
Instead of showing all validation error messages on top, this commit will show each validation error message on the top of respective parameter.

kgaikwad committed:
Fixes #17303 - correct position of validation msg for parameters

tbrisker commented:
Ok to test

orrabin commented:
The error is now in red and above the parameter that is actually causing it as requested:
![image](https://cloud.githubusercontent.com/assets/7894647/20247651/e0d124b8-a9d9-11e6-9fc4-bbf1b4d21998.png)

Maybe stretch the error on the whole row so it doesn't push down each row so much?

Also, please change the commit message to the current status and not the bug, for example it can be: `Fixes #17303 - correct position of validation msg for parameters`

kgaikwad commented:
@orrabin, I have made suggested code changes. Added "colspan" attribute to table row cell.
Please have a look.

orrabin commented on app/assets/javascripts/application.js:
```diff
@@ -154,7 +154,10 @@ function check_caps_lock(key, e) {
-  $(link).closest(".fields").hide();
+  var $field_row = $(link).closest(".fields"),
+  $error_block = $field_row.prev("tr.error-msg-block");
+  if ($error_block) { $error_block.hide(); }
```
Nitpick: no need for the if, calling hide on an empty error_block won't cause an error.

orrabin commented:
@kgaikwad I tested again and it looks much better when stretched, thanks for the change.
Comparing this to the other errors in tables I noticed 2 things:
1) We don't remove the error when clicking remove :+1:  for adding that, maybe a separate pr can remove it in other places.
2) The rest of the tables put the error message under the row that causes it and not above.
Can you try moving the error under the row?

kgaikwad commented:
@orrabin , done with the changes.
As I have removed top border of all rows for better UI with error indication, added little padding on top of each field row and removed the top margin of error row.

kgaikwad commented:
[Test]

orrabin commented:
[test foreman]

orrabin commented:
@Rohoover this pr fixes the location of validation errors in the parameters table, they used to be above the table:
![image](https://cloud.githubusercontent.com/assets/7894647/20386242/678fb3de-acc3-11e6-951a-a880c9e0ef4f.png)

After this pr they are located correctly in the table:
![image](https://cloud.githubusercontent.com/assets/7894647/20386143/f33c9254-acc2-11e6-8475-a54a8262c0b7.png)

but the table doesn't have borders to avoid borders between the parameter and the error (avoid the table below):
![image](https://cloud.githubusercontent.com/assets/7894647/20386360/f5886424-acc3-11e6-8d90-bfcc1a08989b.png)

is it better to keep the table without any borders? with the borders always (even when there is an error)? should it look different?

Rohoover commented:
@orrabin

I don't think the table borders (lines) are necessary as the fields themselves clearly define the rows. If borders are required, I would not put the error in its own row with borders - that visually separates the error from the field and could be confusing. 

Thanks!

orrabin commented:
Thanks @kgaikwad!
